### PR TITLE
Image Classification API Tutorial

### PR DESCRIPTION
**Pull request by luisquintanilla**
Image Classification API Tutorial

[Preview Link](https://review.docs.microsoft.com/en-us/dotnet/machine-learning/tutorials/image-classification-api-transfer-learning?branch=pr-en-us-14722)

[AB#1579022](https://dev.azure.com/mseng/677da0fb-b067-4f77-b89b-f32c12bb8617/_workitems/edit/1579022)

### CODE REVIEW

File: docs/machine-learning/tutorials/image-classification-api-transfer-learning.md [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+The Image Classification API starts the training process by loading a pre-trained TensorFlow model. The training process consists of two steps:
+
+1. Bottleneck phase
+2. Training phase
+
+![](./media/image-classification-api-transfer-learning/training.png)
+
+#### Bottleneck Phase
+
+During the bottleneck phase, the set of training images is loaded and transformed into a format expected by the pre-trained model. These transformations include resizing the image and extracting the pixels into a tensor. A tensor can be thought of as a container that stores data in N-dimensions. Once pre-processed, the data is used as input, or features, for the frozen layers of the model. The frozen layers include all of the layers in the neural network, except the penultimate (bottleneck) layer. These layers are referred to as frozen because no training will occur on these layers and operations are pass-through. The larger the number of layers, the more computationally intensive this step is. Fortunately, since this is a one-time calculation, the results can be cached and used in later runs when experimenting with different parameters.
```
Comment (natke) [inline docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: It's not clear to me why this is called the bottleneck phase, when the bottleneck is the penultimate layer

File: docs/machine-learning/tutorials/image-classification-api-transfer-learning.md [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+- Medical diagnosis
+- Landmark detection
+
+This tutorial trains a custom image classification model to perform automated visual inspection of bridge decks to identify structures that are damaged by cracks.
+
+## ML.NET Image Classification API
+
+ML.NET provides various ways of performing image classification. This tutorial focuses on the Image Classification API. The Image Classification API makes use of [TensorFlow.NET](https://github.com/SciSharp/TensorFlow.NET), a low-level library that provides C# bindings for the TensorFlow C++ API.
+
+![](./media/image-classification-api-transfer-learning/architecture.png)
```
Comment (natke) [inline docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: Do you think an ML.NET user needs to know this level of detail?

File: docs/machine-learning/tutorials/image-classification-api-transfer-learning.md [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+    1. Repeat these steps for **Microsoft.ML.Dnn 0.16.0-preview** and **Microsoft.ML.ImageAnalytics 1.4.0-preview**.
+
+### Prepare and understand the data
+
+> [!NOTE]
+> The datasets for this tutorial are from Maguire, Marc; Dorafshan, Sattar; and Thomas, Robert J., "SDNET2018: A concrete crack image dataset for machine learning applications" (2018). Browse all Datasets. Paper 48. https://digitalcommons.usu.edu/all_datasets/48
+
+SDNET2018 is an image dataset that contains annotations for cracked and non-cracked concrete structures (bridge decks, walls, and pavement). 
+
+![](./media/image-classification-api-transfer-learning/crackeddeck.jpg)
```
Comment (natke) [inline docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: It would be good to have a cracked and non cracked image

File: docs/machine-learning/tutorials/image-classification-api-transfer-learning.md [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+    IDataView imageData = mlContext.Data.LoadFromEnumerable(images);
+    ```
+
+1. The way in which the data is loaded is in the order it was read from the directories. To balance the data, shuffle it using the [`ShuffleRows`](xref:Microsoft.ML.DataOperationsCatalog.ShuffleRows*) method.
+
+    ```csharp
+    IDataView shuffledData = mlContext.Data.ShuffleRows(imageData);
+    ```
+
+1. It's good practice to split the data into a trianing and validation set. The model is trained on the training set of data and how well it makes predictions on unseen data is measured by the performance against the validation set. Split your data
```
Comment (natke) [inline docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: trianing --> training

Comment (natke) [inline docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: I would say this in stronger terms: it's essential to have a validation set, and you can split the training data into a training and validation set, or you can get a validation set from somewhere else

File: docs/machine-learning/tutorials/image-classification-api-transfer-learning.md [hunk 5]
```diff
@@ -0,0 +1,456 @@
...
+    var mapLabelEstimator = mlContext.Transforms.Conversion.MapValueToKey
+        (outputColumnName: "LabelAsKey",
+            inputColumnName: "Label",
+            keyOrdinality: ValueToKeyMappingEstimator.KeyOrdinality.ByValue);
+    ```
+
+1. Then, apply the validation set to the `mapLabelEstimator` using the [`Fit`](xref:Microsoft.ML.ConversionsExtensionsCatalog.MapValueToKey*) method, followed by the [`Transform`](xref:Microsoft.ML.ITransformer.Transform*) method to get the transformed [`IDataView`](xref:Microsoft.ML.IDataView).
+
+    ```csharp
+    IDataView transformedValidationData = mapLabelEstimator
```
Comment (natke) [inline docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: Should this be trainingSet?

**Review by natke — COMMENTED** [review]
Looking good! I think we have a bit more work to do in defining how this is transfer learning, and what bottleneck means.

File: docs/machine-learning/tutorials/image-classification-api-transfer-learning.md (hunk 1 above)
Comment (luisquintanilla) [reply docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: #resolved

File: docs/machine-learning/tutorials/image-classification-api-transfer-learning.md (hunk 3 above)
Comment (luisquintanilla) [reply docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: #resolved

File: docs/machine-learning/tutorials/image-classification-api-transfer-learning.md (hunk 4 above)
Comment (luisquintanilla) [reply docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: #resolved

Comment (luisquintanilla) [reply docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: #resolved

File: docs/machine-learning/tutorials/image-classification-api-transfer-learning.md (hunk 5 above)
Comment (luisquintanilla) [reply docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: #resolved

File: docs/machine-learning/tutorials/image-classification-api-transfer-learning.md [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+ms.author: luquinta
+ms.date: 10/10/2019
+ms.topic: tutorial
+ms.custom: mvc
+#Customer intent: As a developer, I want to use ML.NET so that I can use transfer learning in an image classification scenario to classify images using a pretrained TensorFlow model and ML.NET's Image Classification API.
+---
+
+# Tutorial: Automated visual inspection using transfer learning with ML.NET's Image Classification API
+
+Learn to train a custom deep learning model using transfer learning and the ML.NET Image Classification API to classify images of concrete surfaces as cracked or uncracked.
```
Comment (natke) [inline docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: Learn how to

File: docs/machine-learning/tutorials/image-classification-api-transfer-learning.md [hunk 7]
```diff
@@ -0,0 +1,451 @@
...
+- Medical diagnosis
+- Landmark detection
+
+This tutorial trains a custom image classification model to perform automated visual inspection of bridge decks to identify structures that are damaged by cracks.
+
+## ML.NET Image Classification API
+
+ML.NET provides various ways of performing image classification. This tutorial focuses on the Image Classification API. The Image Classification API makes use of [TensorFlow.NET](https://github.com/SciSharp/TensorFlow.NET), a low-level library that provides C# bindings for the TensorFlow C++ API.
+
+![ML.NET Image Classification API Architecture](./media/image-classification-api-transfer-learning/architecture.png)
```
Comment (natke) [inline docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: This digram looks huge on my mac laptop. Possible to scale?

File: docs/machine-learning/tutorials/image-classification-api-transfer-learning.md [hunk 8]
```diff
@@ -0,0 +1,451 @@
...
+The Image Classification API starts the training process by loading a pre-trained TensorFlow model. The training process consists of two steps:
+
+1. Bottleneck phase
+2. Training phase
+
+![Training Steps](./media/image-classification-api-transfer-learning/training.png)
+
+#### Bottleneck phase
+
+During the bottleneck phase, the set of training images are loaded and the values are used as as input, or features, for the frozen layers of the pre-trained model. The frozen layers include all of the layers in the neural network up to the penultimate layer, informally known as the bottleneck layer. These layers are referred to as frozen because no training will occur on these layers and operations are pass-through. It's at these frozen layers that the lower-level patterns that help a model differentiate between the different classes are computed. The larger the number of layers, the more computationally intensive this step is. Fortunately, since this is a one-time calculation, the results can be cached and used in later runs when experimenting with different parameters.
```
Comment (natke) [inline docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: This is much better!

File: docs/machine-learning/tutorials/image-classification-api-transfer-learning.md [hunk 9]
```diff
@@ -0,0 +1,451 @@
...
+
+#### Training phase
+
+Once the bottleneck values are computed, they are used as input to retrain the final layer of the model. This process is iterative and runs for the number of times specified by model parameters. During each run, the loss and accuracy are evaluated and the appropriate adjustments are made to improve the model with the goal of minimizing the loss and maximizing the accuracy. Once training is finished, two model formats are output. One of them is the `.pb` version of the model and the other is the `.zip` ML.NET serialized version of the model. When working in environemnts supported by ML.NET, it is recommended to use the `.zip` version of the model. However, in environemnts where ML.NET is not supported, you have the option of using the `.pb` version.
```
Comment (natke) [inline docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: What are the bottleneck values?

File: docs/machine-learning/tutorials/image-classification-api-transfer-learning.md (hunk 8 above)
Comment (natke) [inline docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: Although can we be crisper about what exactly is a feature? By this definition the pixels are features?

File: docs/machine-learning/tutorials/image-classification-api-transfer-learning.md [hunk 10]
```diff
@@ -0,0 +1,451 @@
...
+
+## Understand the pretrained model
+
```
Comment (natke) [inline docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: Should we spell out that you could classify many different classes using this model? (Feel free to steal the text from the other transfer learning tutorial (as that will probably be retired once this goes live)

File: docs/machine-learning/tutorials/image-classification-api-transfer-learning.md [hunk 11]
```diff
@@ -0,0 +1,451 @@
...
+1. Use the [`Fit`](xref:Microsoft.ML.Data.EstimatorChain%601.Fit*) method to apply the data to the `preprocessingPipeline` [`EstimatorChain`](xref:Microsoft.ML.Data.EstimatorChain%601) followed by the [`Transform`](xref:Microsoft.ML.Data.TransformerChain`1.Transform*) method which returns an [`IDataView`](xref:Microsoft.ML.IDataView) containing the pre-processed data.
+
+    ```csharp
+    IDataView preProcessedData = preprocessingPipeline
+                                    .Fit(shuffledData)
+                                    .Transform(shuffledData);
+    ```
+
+1. To train a model, it's important to have a training dataset as well as a validation dataset. The model is trained on the training set of data and how well it makes predictions on unseen data is measured by the performance against the validation set. Based on the results of that performance, the model makes adustments to what it has learned in an effort to improve. The validation set can come from either splitting your original dataset or from another source that has already been set aside for this purpose. In this case, the pre-processed dataset is split into training, validation and test sets.
```
Comment (natke) [inline docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: Suggestion:

To train a model, it's important to have a training dataset as well as a validation dataset. The model is trained on the training set. How well it makes predictions on unseen data is measured by the performance against the validation set.

Comment (natke) [inline docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: adustments -> adjustments

File: docs/machine-learning/tutorials/image-classification-api-transfer-learning.md [hunk 12]
```diff
@@ -0,0 +1,451 @@
...
+    ```
+
+    The code sample above performs two splits. First, the pre-processed data is split and 70% is used for training while the remaining 30% is used for validation. Then, the 30% validation set is further split into validation and test sets where 90% is used for validation and 10% is used for testing. 
+
+    A way to think about the purpose of these data partitions is taking an exam. When studying for an exam, you review your notes, books or other resources to get get a grasp on the concepts that are on the exam. This is what the train set is for. Then, you might take a mock exam to validate your knowledge. This is where the validation set comes in handy. You want to check whether you have a good grasp of the concepts before taking the actual exanm. Based on those results, you take note of what you got wrong or didn't understand well and incorporate your changes as you review for the real exam. Finally, you take the exam. This is what the test set is used for. You've never seen the questions that are on the exam and now use what you learned from training and validation to apply your knowledge to the task at hand.
```
Comment (natke) [inline docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: exanm --> exam

**Review by natke — COMMENTED** [review]
A few more comments.

File: docs/machine-learning/tutorials/image-classification-api-transfer-learning.md (hunk 12 above)
Comment (luisquintanilla) [reply docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: #resolved

File: docs/machine-learning/tutorials/image-classification-api-transfer-learning.md (hunk 11 above)
Comment (luisquintanilla) [reply docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: #resolved

Comment (luisquintanilla) [reply docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: #resolved

File: docs/machine-learning/tutorials/image-classification-api-transfer-learning.md (hunk 9 above)
Comment (luisquintanilla) [reply docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: #resolved

File: docs/machine-learning/tutorials/image-classification-api-transfer-learning.md (hunk 8 above)
Comment (luisquintanilla) [reply docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: #resolved

File: docs/machine-learning/tutorials/image-classification-api-transfer-learning.md (hunk 10 above)
Comment (luisquintanilla) [reply docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: #resolved

File: docs/machine-learning/tutorials/image-classification-api-transfer-learning.md (hunk 2 above)
Comment (luisquintanilla) [reply docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: #resolved

File: docs/machine-learning/tutorials/image-classification-api-transfer-learning.md (hunk 7 above)
Comment (luisquintanilla) [reply docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: #resolved

**Review by cjgronlund — CHANGES_REQUESTED** [review]
Spelling error
image

**Review by cjgronlund — CHANGES_REQUESTED** [review]
One more: 
image

**Review by cjgronlund — CHANGES_REQUESTED** [review]
Need a period or colon here:
image

**Review by cjgronlund — CHANGES_REQUESTED** [review]
Need a period here:
image

**Review by cjgronlund — CHANGES_REQUESTED** [review]
Capital S: Search - 
image

**Review by cjgronlund — COMMENTED** [review]
Should using be in code style here? 
image

**Review by cjgronlund — CHANGES_REQUESTED** [review]
Change to: formatted
Also, should "list if" be "list of"?
image

**Review by cjgronlund — CHANGES_REQUESTED** [review]
Needs a period or colon:
image

**Review by cjgronlund — CHANGES_REQUESTED** [review]
Needs a period at the end:
image

**Review by cjgronlund — CHANGES_REQUESTED** [review]
Make this a period, not a comma:
image

**Review by cjgronlund — CHANGES_REQUESTED** [review]
This sentence is hard to parse. Make it simpler?
The data is loaded in the order it was read from the directories. 
image

**Review by cjgronlund — COMMENTED** [review]
Most prefixes **don't** take a hyphen: 
preprocessing (not pre-processing)
pretrained (not pre-trained)
subdirectory

**Review by cjgronlund — COMMENTED** [review]
For later: This list can link to the descriptions in the API ref when it's published. We should duplicate API ref material (or material that **should be** in the ref) here.
image

**Review by cjgronlund — CHANGES_REQUESTED** [review]
"and" wanted here?
image

File: docs/machine-learning/tutorials/image-classification-api-transfer-learning.md [hunk 13]
```diff
@@ -0,0 +1,451 @@
...
+
+#### Training phase
+
+Once the output values from the bottleneck phase are computed, they are used as input to retrain the final layer of the model. This process is iterative and runs for the number of times specified by model parameters. During each run, the loss and accuracy are evaluated and the appropriate adjustments are made to improve the model with the goal of minimizing the loss and maximizing the accuracy. Once training is finished, two model formats are output. One of them is the `.pb` version of the model and the other is the `.zip` ML.NET serialized version of the model. When working in environemnts supported by ML.NET, it is recommended to use the `.zip` version of the model. However, in environemnts where ML.NET is not supported, you have the option of using the `.pb` version.
```
Comment (mairaw) [inline docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: @cjgronlund it's easier to submit corrections via suggestions and in bulk to keep our inboxes sane. Did an example below.

image


```suggestion
Once the output values from the bottleneck phase are computed, they are used as input to retrain the final layer of the model. This process is iterative and runs for the number of times specified by model parameters. During each run, the loss and accuracy are evaluated and the appropriate adjustments are made to improve the model with the goal of minimizing the loss and maximizing the accuracy. Once training is finished, two model formats are output. One of them is the `.pb` version of the model and the other is the `.zip` ML.NET serialized version of the model. When working in environments supported by ML.NET, it is recommended to use the `.zip` version of the model. However, in environments where ML.NET is not supported, you have the option of using the `.pb` version.
```

**Review by cjgronlund — CHANGES_REQUESTED** [review]
Other changes:

- Diagram under ML.NET Image Classification API needs more detail to be helpful. Remove for now?
- "What is transfer learning?" belongs in its own article. For now, you can bump up the heading levels by one, so that "What is...?" is an H2. Don't use heading levels below H3 in most cases.
- Need x-refs under Next steps. Are there related ML.NET concepts or tutorials you can link to? Link to "Deep Learning" article? https://docs.microsoft.com/en-us/azure/machine-learning/service/concept-deep-learning-vs-machine-learning
- You don't need to repeat the checklist at the end of the article. You can sum it up in a sentence, if you want.

File: docs/machine-learning/tutorials/image-classification-api-transfer-learning.md (hunk 13 above)
Comment (cjgronlund) [reply docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: Thanks for the tip. Sorry for the spam!

Comment (mairaw) [reply docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: No worries. Suggestions is an awesome feature and then we can later apply them in bulk as well.

**Comment by luisquintanilla** [comment]
@cjgronlund thanks for the review. I think I've addressed all the comments.

**Comment by luisquintanilla** [comment]
@cjgronlund I've made updates to take into account the latest sample. Ready for review.

File: docs/machine-learning/tutorials/image-classification-api-transfer-learning.md [hunk 14]
```diff
@@ -0,0 +1,529 @@
...
+    - `reuseTrainSetBottleneckCachedValues` tells the model whether to use the cached values from the bottleneck phase in subsequent runs. The bottleneck phase is a one-time pass-through computation that is computationally intensive the first time it is performed. If the training data does not change and you want to experiment using a different number of epochs or batch size, using the cached values significantly reduces the amount of time required to train a model.
+    - `reuseValidationSetBottleneckCachedValues` is similar to `reuseTrainSetBottleneckCachedValues` only that in this case it's for the validation dataset.
+    - `disableEarlyStopping` tells the Image Classification whether to employ an early stopping strategy. As the model searches for the optimal values that will help it make accurate predictions during training, performance may increase or decrease. Ultimately, if the model reaches the last epoch, it may be the case that the patterns it learned from training are suboptimal. Early stopping monitors training for these drops in performance and stops the training process in an effort to preserve an optimal version of the model.
```
Comment (cjgronlund) [inline docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]:
```suggestion
    - `disableEarlyStopping` tells the ImageClassification estimator whether to employ an early stopping strategy. As the model searches for the optimal values that will help it make accurate predictions during training, performance may increase or decrease. Ultimately, if the model reaches the last epoch, it may be the case that the patterns it learned from training are suboptimal. Early stopping monitors training for these drops in performance and stops the training process in an effort to preserve an optimal version of the model.
```

Comment (cjgronlund) [reply docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: Instead of "the Image Classification" did you mean "the ImageClassification estimator" that this list is about?

**Review by cjgronlund — COMMENTED** [review]
A few nits, but overall an excellent tutorial, @luisquintanilla!

**Comment by luisquintanilla** [comment]
Retriggering build

**Comment by lobrien** [comment]
ping.

**Comment by luisquintanilla** [comment]
@lobrien Pong

File: docs/machine-learning/tutorials/image-classification-api-transfer-learning.md [hunk 15]
```diff
@@ -0,0 +1,399 @@
...
+ms.author: luquinta
+ms.date: 10/25/2019
+ms.topic: tutorial
+ms.custom: mvc
+#Customer intent: As a developer, I want to use ML.NET so that I can use transfer learning in an image classification scenario to classify images using a pretrained TensorFlow model and ML.NET's Image Classification API.
+---
+
+# Tutorial: Automated visual inspection using transfer learning with the ML.NET Image Classification API
+
+Learn how to train a custom deep learning model using transfer learning, a pretrained TensorFlow model and the ML.NET Image Classification API to classify images of concrete surfaces as cracked or uncracked.
```
Comment (cjgronlund) [inline docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]:
```suggestion
Learn how to train a custom deep learning model using transfer learning, a pretrained TensorFlow model and the [ML.NET Image Classification API](URL?) to classify images of concrete surfaces as cracked or uncracked.
```

Comment (cjgronlund) [reply docs/machine-learning/tutorials/image-classification-api-transfer-learning.md]: Provide a link for the Image Classification API? It sounds like an API separate from the ML.NET API, but I don't know where to find it.

**Review by cjgronlund — COMMENTED** [review]
Looks good! Thank you, Luis!

**Review by cjgronlund — APPROVED** [review]
